### PR TITLE
Refactoring common HAL host code to enable marl integration.

### PR DESCRIPTION
**Pull request by benvanik**
This will let us toggle between the existing scheduling model (simple serial/in-order on a single worker thread) to marl (out-of-order on a pool of workers). Keeping the existing serial model is useful for systems where marl may not be supported yet (wasm, esoteric embedded devices, etc) and should also be easier to debug (as no fibers confusing debuggers).

Progress on #49.

### CODE REVIEW

File: iree/hal/dylib/dylib_executable.cc [hunk 1]
```diff
@@ -81,17 +94,57 @@ Status DyLibExecutable::Initialize() {
...
+      dispatch_state->descriptors.push_back(descriptor);
+      dispatch_state->args.push_back(&descriptor->descriptor);
+    }
   }
+
+  return std::move(dispatch_state);
 }
 
-Status DyLibExecutable::Invoke(int func_id, absl::Span<void*> args) const {
-  return UnimplementedErrorBuilder(IREE_LOC) << "DyLibExecutable::Invoke NYI";
```
Comment (ScottTodd) [inline iree/hal/dylib/dylib_executable.cc:94]: :D thanks for implementing

File: iree/hal/host/scheduling_model.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+#include "iree/hal/command_queue.h"
+
+namespace iree {
+namespace hal {
+
+// Host-local scheduling interface that device implementations can use to choose
+// between various scheduling strategies (such as serial/in-order,
+// fiber/out-of-order, etc). The interface models a subset of the Device
+// interface relating to the scheduling primitives (such as semaphores) and the
+// device-level operations that can be performed on them (such as wait-all).
```
Comment (ScottTodd) [inline iree/hal/host/scheduling_model.h:27]: As always, great comments on this class and its methods :D thanks!

File: iree/hal/vmla/vmla_executable.cc [hunk 3]
```diff
@@ -105,27 +108,81 @@ Status VMLAExecutable::Initialize(iree_vm_instance_t* instance,
...
+  IREE_TRACE_SCOPE0("VMLAExecutable::PrepareDispatch");
+
+  if (params.entry_point >= entry_functions_.size()) {
+    return InvalidArgumentErrorBuilder(IREE_LOC)
+           << "Invalid entry point ordinal " << params.entry_point;
+  }
+
+  auto dispatch_state = make_ref<VMLADispatchState>();
+  dispatch_state->function = entry_functions_[params.entry_point];
+  dispatch_state->input_list_size = iree_vm_variant_list_alloc_size(1 + 3);
```
Comment (ScottTodd) [inline iree/hal/vmla/vmla_executable.cc:136]: document what 1 and 3 are here?

File: iree/hal/vmla/vmla_executable.cc [hunk 4]
```diff
@@ -105,27 +108,81 @@ Status VMLAExecutable::Initialize(iree_vm_instance_t* instance,
...
+}
+
+Status VMLAExecutable::DispatchTile(DispatchState* state,
+                                    std::array<uint32_t, 3> workgroup_xyz) {
+  IREE_TRACE_SCOPE0("VMLAExecutable::DispatchTile");
+  auto* dispatch_state = static_cast<VMLADispatchState*>(state);
+
+  auto* input_list = reinterpret_cast<iree_vm_variant_list_t*>(
+      alloca(dispatch_state->input_list_size));
+  iree_vm_variant_list_init(input_list, 1 + 3);
```
Comment (ScottTodd) [inline iree/hal/vmla/vmla_executable.cc:167]: document 1 and 3 here too?

File: SUBMODULE_VERSIONS [hunk 5]
```diff
@@ -4,6 +4,7 @@ daff5fead3fbe22c6fc58310ca3f49caf117f185 third_party/benchmark
 a5d9d0f7d368054fd1691aedf1db4116efcc233e third_party/flatbuffers
 f2fb48c3b3d79a75a88a99fba6576b25d42ec528 third_party/googletest
 755a89591528b692315ad0325347e2fd4637271b third_party/llvm-project
+17b12a4481daa150e2d1ea3ada086b551b856707 third_party/marl
```
Comment (ScottTodd) [inline SUBMODULE_VERSIONS:7]: Technically adding the marl submodule doesn't need to happen at the same time as these other changes, but I'm fine with adding it in this PR.

**Review by ScottTodd — APPROVED** [review]
Looks like there are some conflicts to resolve and a test to disable/fix (as discussed elsewhere):
```
//iree/test/e2e/models:edge_detection.mlir.test                         TIMEOUT in 71.8s
```

**Comment by phoenix-meadowlark** [comment]
I'm restructuring some parts of the CI, and all open PRs behind [this commit](https://github.com/google/iree/commit/f3000d04b151fe87614083b30c131329225bf4ec) will be affected. This only affects PRs that are seven days old or older. Expect presubmit failures for `linux_bazel_core` and eventually `linux_bazel` on future pushes until this PR is synced.

**Pull request closed without merging**